### PR TITLE
Search autocompletion.

### LINKED ISSUE #79: Search engine dynamic suggestions (alias "webjumps")
I think it's important to discuss the bookmark format at an early stage.

1. A binary format is probably not a good idea.  It's not versionable and you are going to lose a huge user base who likes to keep their bookmarks in a git repo.
A few years back, another power-browser transited from text format to SQLite, it went defunct in the few following months... (I think it was Luakit.)  Also see https://github.com/qutebrowser/qutebrowser/issues/882#issuecomment-136807011.

2. What do you think about using Emacs `eww`'s format?  That would allow the two to share the same bookmarks transparently, a big bonus for Emacs users I think.

3. Qutebrowser is currently discussing the merge of bookmarks, quickmarks and search engines.  I think it's a neat idea.  Basically every bookmark can be recalled with a simple keybind (if specified) or used as a search engine (again, if specified).
See:
https://github.com/qutebrowser/qutebrowser/issues/882.

### PR DESCRIPTION
**Pull request by aartaka**
Behold! Search suggestions in Nyxt! (see an attached screenshot)

autocompletion

This adds search autocompletion requests to the default `search-engine`s, based on the new extensible mechanism:
- `completion-function` slot of `search-engine', and
- `make-search-completion-function` helper to make completion functions.

Works, loads things (although takes some time with both Wikipedia and DuckDuckGo), shows fancy search suggestions!

(I'm going to add completion functions to `nx-search-engines` right after we merge this, as this is too exciting to ignore!)

How does it look for you?

Closes #79.

EDIT: Small readability fixes.
EDIT: "deafult" typo :)
EDIT: Add a linked issue.

### CODE REVIEW

**Comment by jmercouris** [comment]
Absolutely blows my mind! Fantastic work Artyom!

That is just so damn cool!

**Comment by Ambrevar** [comment]
Awesoooooooooooooooooooome!!! Thank you so much for working on this, will review as soon as possible!

P.S.: I see you had fun with theming :)

**Comment by jmercouris** [comment]
It seems Pierre is not familiar with the venerable Laconia theme!

**Comment by Ambrevar** [comment]
The related issue is #79, one of our oldest! ;)

**Comment by jmercouris** [comment]
Resources: https://github.com/aartaka/nyxt-config/blob/master/style.lisp
https://github.com/aartaka/laconia-theme

**Comment by Ambrevar** [comment]
I've tested it, it's absolutely amazing.
I suggest we merge this now and include it in 2-pre-release-7.
@jmercouris ?

File: source/search-engine.lisp [hunk 1]
```diff
@@ -32,6 +36,28 @@ the empty string."))
...
                  :fallback-url fallback-url))
 
+(export-always 'make-search-completion-function)
+(declaim (ftype (function (&key (:base-url string)
+                                (:request-function (function (string &rest *) *))
+                                (:processing-function (function (*) list-of-strings)))
+                          (function (string) list-of-strings))
+                make-search-completion-function))
+(defun make-search-completion-function (&key base-url
+                                          (request-function #'dex:get)
```
Comment (Ambrevar) [inline source/search-engine.lisp:46]: dex:get must use some proxy settings, no?
But which ones?  Does it make sense for the prompt buffer to use proxy settings of the current-buffer?
Should it have its own proxy settings?

Here is my suggestion: make a wrapper around `dex:get` that's uses the prompt buffer proxy.
Then the user can enable the proxy by adding proxy-mode to the prompt-buffer or not.
Does that make sense?

**Comment by jmercouris** [comment]
I think we best leave it for actual 2.0.0. It is too much of a last minute thing in my opinion. Rather, let Artyom fix it in peace :-). What do you think?

**Comment by Ambrevar** [comment]
Agreed!

**Comment by aartaka** [comment]
Things left to do before merging:
- auto-completion using default search engine, e.g., when you type `salad` and it gives you `salad recipe`, `salad dressing` etc. even without the search engine shortcut.
- Proxying.
- Making auto-completion togglable: there are people as paranoidal as myself in regards to auto-completion, so we better make it disableable :D

**Comment by Ambrevar** [comment]
Agreed on all points! :)

**Comment by aartaka** [comment]
- Implemented:
  - Proxied auto-completion.
  - A buffer-located switch for auto-completion.
  - Auto-completion by default, for any query.
- Rebased over master to use the fix for the bug found in the process (11425b2300cbe295fc25beef7c52ebaf6a67cc20).
- Tested proxied downloads and disabled auto-completion. Works fine.

Mergeable, I believe :)

**Comment by aartaka** [comment]
+ CI passes. Phew...

File: source/search-engine.lisp [hunk 2]
```diff
@@ -32,6 +36,38 @@ the empty string."))
...
 
+;; TODO: Move elsewhere?
+;; TODO: Add `proxied-post'?
+(defun proxied-get (url &rest args)
+  "A version of `dex:get' that enables proxy in case `current-buffer' has one enabled."
+  (let ((args (append args (list :proxy (and (web-buffer-p (current-buffer))
+                                             (proxy-address (current-buffer))
+                                             (quri:render-uri
+                                              (proxy-address (current-buffer))))))))
+    (apply #'dex:get url args)))
```
Comment (Ambrevar) [inline source/search-engine.lisp:47]: What about adding a `(proxy-url-string BUFFER)` function, then at the call site use

```lisp
(dex:get :proxy (proxy-url-string PROMPT-BUFFER))
```

?

By the way, shouldn't we rename `proxy-address` to `proxy-url` for consistency?

File: source/buffer.lisp [hunk 3]
```diff
@@ -992,17 +1006,30 @@ Finally, if nothing else, set the `engine' to the `default-search-engine'."))
...
-                   (all-search-engines)
-                   :key #'shortcut)))
-    (append (unless (and engines (string= (first terms) (shortcut (first engines))))
-              (list (make-instance 'new-url-query :query input)))
-            (mapcar (lambda (engine) (make-instance 'new-url-query
-                                                    :query (str:join " " (rest terms))
-                                                    :engine engine))
-                    engines))))
+  (flet ((make-engine-completions (engine input)
+           (when (and (and (current-buffer) (search-auto-complete-p (current-buffer)))
```
Comment (Ambrevar) [inline source/buffer.lisp:1010]: Nested `and` can be merged.

**Comment by Ambrevar** [comment]
So how does one configure the prompt-buffer to be proxied?

**Comment by aartaka** [comment]
There's no way as of now: I use `current-buffer` (which prompt-buffer cannot be) proxy to decide on proxying auto-completion. So, prompt-buffer is not involved. 

Should prompt-buffer have a proxy? I'm not sure -- prompt-buffer is made for internal interface use, after all.

File: source/search-engine.lisp (hunk 2 above)
Comment (aartaka) [reply source/search-engine.lisp:47]: Hmmm. I'm not sure about the `proxy-url-string` helper. Let's leave it the way it is and see whether we need it in the future :)

Renaming `proxy-adress` to `proxy-url` sounds intuitive! I'll do that.

**Comment by jmercouris** [comment]
I agree with Artyoms assessment

**Comment by jmercouris** [comment]
I understood that the prompt buffer will use the proxy settings of the current buffer

**Comment by Ambrevar** [comment]
> I understood that the prompt buffer will use the proxy settings of the current buffer

But that's not correct.  For instance it will fail to do the right thing
when run from the *Help* buffer.

**Comment by aartaka** [comment]
Fair point! While I don't feel right about adding proxy to the prompt-buffer, I see the problem.

**Comment by jmercouris** [comment]
I would say don't add it to the prompt buffer, add it to the browser, that makes more sense. It can then be used for downloads with the lisp engine etc.

**Comment by jmercouris** [comment]
I don't see that as likely, but a fair point. Being more specific has benefits.

It also has costs. Will the user know to set their proxy for every single aspect of the system?

**Comment by Ambrevar** [comment]
> It also has costs. Will the user know to set their proxy for every single aspect of the system?

We can set the default value to that of (proxy *browser*).

Done! :)

**Comment by Ambrevar** [comment]
Sorry, I wasn't very clear: We can set the default value _of the various local proxy slots_ to that of `(proxy *browser*)`.

**Comment by jmercouris** [comment]
I see, that seems a good solution.

**Comment by aartaka** [comment]
Hmmmm... How about yet another approach:
- Don't define auto-completion functions in Nyxt core.
- Leave proxied completion functions to be implemented by extension writers (because this logic actually belongs to extensions).
- Add `:request-args` keyword argument to the `make-search-completion-function` and use `dex:get` by default for `:request-function`. Example of usage:
```lisp
(make-search-completion-function
 :base-url "https://duckduckgo.com/ac/?q=~a"
 :request-args '(:proxy "socks5://localhost:9050")
 :processing-function
 #'(lambda (results)
     (mapcar #'cdar
             (json:decode-json-from-string results))))
```

Given that `nx-search-engines` is already destined to be a testbed for completion, I suggest completions functions (including proxied ones) are written there.

Rationale: 
- `prompt-buffer` is merely a display for search suggestions, so it doesn't make sense to add proxies to it.
- `buffer` makes more sense, given that most search-related slots (`search-engines`, `search-auto-complete-p`) are there. Why not use `buffer`'s `proxy` then? Because it's not every buffer that has proxies. Maybe instantiate buffers and enable `proxy-mode`? That's unreliable -- too much variables. Adding search-dedicated proxy slot doesn't sound right either.
- Adding `*browser*`-based proxy and inheriting that configuration is cool, but that's another global variable that supercedes a carefully crafted mode. Doesn't sound right.
- The mechanism is intended to be extensible, so why impose any implementation? If user wants to have things proxied, they can always override the `completion-function`. `make-search-completion-function` is intended for that, especially so with `:request-args`.

**Comment by jmercouris** [comment]
I concur that the `make-search-completion-function` is the appropriate way for the user to customize this. It does not feel right that the `prompt-buffer` contain proxy information. It seems to me that would be outside of the prompt buffers concerns.

**Comment by Ambrevar** [comment]
I agree with your last proposal, @aartaka! :)

**Comment by aartaka** [comment]
> I agree with your last proposal, @aartaka! :)

I'll push it in an hour then!

**Comment by aartaka** [comment]
Here it is: `:request-args` and `dex:get` used by default. Two things to discuss:
- Should I remove the default search engines' (DDG, Wiki) completion?
- `proxy-url` in `buffer` conflicts with `proxy-url` in `renderer-gtk`. Maybe rename GTK one?

**Comment by jmercouris** [comment]
Thank you Artyom! I would leave the default search engine completion for now. I will look into what to do about the naming conflict!

**Comment by jmercouris** [comment]
Perhaps it would make sense to name it %proxy-url in renderer-gtk.lisp, since it is an implementation detail.

**Comment by jmercouris** [comment]
Or is it...? I guess it isn't. Decisions decisions...

**Comment by jmercouris** [comment]
No, definitely it should be set to %proxy-url in `renderer-gtk.lisp`.

**Comment by jmercouris** [comment]
@aartaka if you approve, I shall merge!

**Comment by jmercouris** [comment]
I decided to just call it `gtk-proxy-url` in the end.

**Comment by Ambrevar** [comment]
Looks good to me, except that I think we should expand add an example on how to add a proxy, and refer to this in the manual.

**Comment by aartaka** [comment]
> Looks good to me, except that I think we should expand add an example on how to add a proxy, and refer to this in the manual.

Fair point. Done in 65092494ed1356baced050a207e76803db99cb7a. Feels all good to merge otherwise :)

**Pull request merged**